### PR TITLE
Migrate to NET 4.5

### PR DESCRIPTION
**Pull request by shatl**
Migrate to .NET 4.5;
Update dependencies (Json.Net 6, ASP.NET MVC 5, NHibernate 4);

NOTE:
1) SharpArchitecture-Build needs to be merged first;
2) Submodule BuildSystem reference needs to be re-pointed to SharpArchitecture repository.

### CODE REVIEW

**Comment by shatl** [comment]
@seif Hi Seif, any chances to get this merged soon?

**Comment by kad0xf** [comment]
Hi shatl,
Why are you trying to commit DLL files in ReferencedAssemblies (e.g. ReferencedAssemblies/ASP.NET MVC/System.Web.Mvc.dll) ?
You should use nuget for MVC 5 references (similar to what I did for MVC 4: https://github.com/sharparchitecture/Sharp-Architecture/commit/723fe628c5b0634652b4ef1d891382cee75a86fc).

**Comment by shatl** [comment]
Hi @kad0xf 
This was done by mistake,
Could you help me cleanup / squash commits?

**Comment by kad0xf** [comment]
@shatl, Never mind my comment. I did not pay attention to the fact that you deleted the DLL files, which is OK. Did you try to build and test the project? I get two errors when I try to run Build/BuildAndTest.cmd. I am trying now to find the cause of these errors.

**Comment by shatl** [comment]
@kad0xf Works for me. Can you post error messages?

**Comment by kad0xf** [comment]
I managed to solve the errors.

**Comment by seif** [comment]
Sorry for the delay looking at this, I'll review it tomorrow. Thanks


Apologies in advance for any auto correction or spelling mistakes, sent
from my phone.

**Comment by kad0xf** [comment]
@shatl I managed to build and publish your fork as NuGet packages. You can download the test the new packages from NuGet: https://www.nuget.org/packages/SharpArchitecture/

**Comment by shatl** [comment]
@kad0xf So we aren't using old package names now?

I'll test by tomorrow

**Comment by shatl** [comment]
@kad0xf  I'm getting this error:
Attempting to resolve dependency 'SharpArchitecture.Domain (≥ 2.4.0 && < 3.0.0)'.
Unable to resolve dependency 'SharpArchitecture.Domain (≥ 2.4.0 && < 3.0.0)'.


Nuget is not able to resolve beta references (even if I have Include Prerelease selected in Manage NuGet packages dialog)

File: NugetTemplates/SharpArch.NHibernate/SharpArch.NHibernate.nuspec [hunk 1]
```diff
@@ -13,9 +13,11 @@
     <summary>Architectural foundation for building maintainable applications with Domain Driven Design patterns and practices</summary>
     <language>en-US</language>
     <dependencies>
-	  <dependency id="SharpArch.Domain" version="(1.9.6,3.0.0)" />
-      <dependency id="FluentNhibernate" version="[1.4,2.0)" />
-      <dependency id="Microsoft.AspNet.Mvc" version="[4.0.20710,4.0.30506.0]" />
+	  <dependency id="SharpArch.Domain" version="[2.4.0,3.0.0)" />
+      <dependency id="FluentNhibernate" version="[2.0.1.0,3.0.0.0)" />
+      <dependency id="Microsoft.AspNet.Mvc" version="[5.2.2,6.0.0)" />
+      <dependency id="NHibernate.Validator" version="[1.3.2.0,2.0.0.0)" />
```
Comment (seif) [inline NugetTemplates/SharpArch.NHibernate/SharpArch.NHibernate.nuspec]: How come we are introducing NHibernate.Validator again? If I remember correctly, we don't use it any more.

File: Solutions/SharpArch.Domain/SharpArch.Domain.csproj [hunk 2]
```diff
@@ -10,11 +10,10 @@
...
     <RootNamespace>SharpArch.Domain</RootNamespace>
     <AssemblyName>SharpArch.Domain</AssemblyName>
-    <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
-    <SignAssembly>true</SignAssembly>
-    <AssemblyOriginatorKeyFile>
-    </AssemblyOriginatorKeyFile>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>..\SharpArch.snk</AssemblyOriginatorKeyFile>
```
Comment (seif) [inline Solutions/SharpArch.Domain/SharpArch.Domain.csproj]: Wouldn't this cause assembly signing to kick in? we don't want to sign the assemblies.

**Comment by seif** [comment]
Nice one @shatl just the use of lamda concurrent dictionary.
I noticed the SignAssembly false property later, so it might not be strong signing the assemblies, but I need to check that.

There is a bunch of commits with pulling from upstream or merging master, any way to cherry pick around those or get rid of them somehow?

For future reference, this pull request had a bunch of unrelated changes, it is better to split them into smaller pull requests (perf improvements/dotsettings file/net45 mvc5). That way we can discuss changes more.

I will probably up the version to 3.0, as there is a breaking change and targeting different framework version.

**Comment by seif** [comment]
@kad0xf if you are going to fork and release nuget packages, you should change the name and specify what is different from official update links to point to your fork. With the name you gave, it is very confusing, can you take it down please.

**Comment by kad0xf** [comment]
@seif no problem, I can take it down. It was meant to be for @shatl and my self to test the recent changes. But when will you update the official packages? You have not even released the changes I made in August last year (when I updated the web project to MVC4)!

File: NugetTemplates/SharpArch.NHibernate/SharpArch.NHibernate.nuspec (hunk 1 above)
Comment (kad0xf) [inline NugetTemplates/SharpArch.NHibernate/SharpArch.NHibernate.nuspec]: I believe that @shatl forked an old version of the project (Feb 04, 2014).

**Comment by shatl** [comment]
@seif Cherry pick does not like me :( 
I'll try to split pull into two: .NET migration and perf. optimization.

**Comment by shatl** [comment]
@seif  Let's just close this one. I'll create two separate requests - one for .NET 4.5 and another one for performance

**Comment by shatl** [comment]
@seif  use this request instead #109

**Pull request closed without merging**